### PR TITLE
[issue#256]Implement SQL++ Query Generator

### PR DESCRIPTION
**Pull request by luochen01**
Issue #256

This PR mainly includes the following changes:

Added an abstract AsterixQueryGenerator, which is the base class for AQL/SQLPP Generators.
Added a class SQLPPGenerator, which implements the functionality of generating SQLPP Queries.
Added test cases SQLPPGeneratorTest, which includes all tests from AQLGeneratorTest.

### CODE REVIEW

**Comment by JavierJia** [comment]
Nice Split! Thanks. This one will be easier to review :-)

**Comment by chenlica** [comment]
Can @waans11 review it as well?

**Comment by JavierJia** [comment]
sure. I added @waans11 as reviewer.

File: zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/AsterixQueryGenerator.scala [hunk 1]
```diff
@@ -0,0 +1,270 @@
+package edu.uci.ics.cloudberry.zion.model.impl
+
+import edu.uci.ics.cloudberry.zion.model.datastore.{IQLGenerator, QueryParsingException}
+import edu.uci.ics.cloudberry.zion.model.schema._
+
+
+/**
+  * Defines constant strings for query languages supported by AsterixDB
+  */
+trait TypeImpl {
```
Comment (JavierJia) [inline zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/AsterixQueryGenerator.scala]: Rename to `AsterixDB` related `impl`?

File: zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/AsterixQueryGenerator.scala [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+
+  def getAggregateStr(aggregate: AggregateFunc): String = {
+    aggregateFuncMap.get(aggregate) match {
+      case Some(impl) => impl
+      case None => throw new QueryParsingException(s"No implementation is provided for aggregate function ${aggregate.name}")
+    }
+  }
+
+  val datetime: String = "datetime"
+  val round: String = "round"
```
Comment (JavierJia) [inline zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/AsterixQueryGenerator.scala]: `trait` should be as clean as possible. we can move this static values to object.

File: zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/AsterixQueryGenerator.scala [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+
+  protected def unnestVar: String
+
+  protected def groupVar: String
+
+  protected def globalAggrVar: String
+
+  protected def outerSelectVar: String
+
+  protected def suffix: String
```
Comment (JavierJia) [inline zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/AsterixQueryGenerator.scala]: what is this `suffix` ? mind to add a scala doc or change the name tobe clear?

File: zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/AsterixQueryGenerator.scala [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+      case _ => {
+        s"$fieldExpr ${filter.relation} ${typeImpl.datetime}('${filter.values(0)}')"
+      }
+    }
+  }
+
+
+  protected def parsePointRelation(filter: FilterStatement,
+                                   fieldExpr: String): String = {
+    val values = filter.values.map(_.asInstanceOf[Seq[Double]])
```
Comment (JavierJia) [inline zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/AsterixQueryGenerator.scala]: this method seems haven't been tested, could you add one?

File: zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/AsterixQueryGenerator.scala [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+    */
+  case class FieldExpr(refExpr: String, defExpr: String)
+
+  /**
+    * Partial parse results after parsing each [[Statement]]
+    *
+    * @param strs    a sequence of parsed query strings, which would be composed together later.
+    * @param exprMap a new field expression map
+    */
+  case class PartialResult(strs: Seq[String], exprMap: Map[String, FieldExpr])
```
Comment (JavierJia) [inline zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/AsterixQueryGenerator.scala]: `PartialResult` has special meaning in the query slicing context, to make it less confusing, we may call it `ParsedResult`?

File: zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/schema/Query.scala [hunk 6]
```diff
@@ -35,6 +35,12 @@ case class Query(dataset: String,
                  isEstimable: Boolean = false
                 ) extends IReadQuery {
 
+  def unnested: Boolean = !unnest.isEmpty
```
Comment (luochen01) [inline zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/schema/Query.scala]: Agreed!

File: zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/AsterixQueryGenerator.scala (hunk 5 above)
Comment (luochen01) [inline zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/AsterixQueryGenerator.scala]: Sure.

File: zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/AsterixQueryGenerator.scala (hunk 3 above)
Comment (luochen01) [inline zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/AsterixQueryGenerator.scala]: Some query language (like SQL++) requires ";" after each query, while AQL does not. Suffix actually represents this ";" appended to each query. Just added some doc to this.

File: zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/AsterixQueryGenerator.scala (hunk 2 above)
Comment (luochen01) [inline zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/AsterixQueryGenerator.scala]: Here AQL and SQL++ actually use the same implementation of datetime and round, thus their implementations are extracted to AsterixImpl.
If we move them to the companion object, then we have to use AsterixImpl.datetime and AsterixImpl.round to use them, which are not compatible with other fields...

File: zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/AsterixQueryGenerator.scala (hunk 4 above)
Comment (luochen01) [inline zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/AsterixQueryGenerator.scala]: Sure, added a test case on this.

File: zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/AsterixQueryGenerator.scala (hunk 2 above)
Comment (JavierJia) [inline zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/AsterixQueryGenerator.scala]: Then I will suggest to remove the implementation and implement the `val` separately in both classes. Though the implementation is exactly the same, it make the `trait` more safe.

File: zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/AsterixQueryGenerator.scala (hunk 1 above)
Comment (luochen01) [inline zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/AsterixQueryGenerator.scala]: Done.

File: zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/AsterixQueryGenerator.scala (hunk 2 above)
Comment (luochen01) [inline zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/AsterixQueryGenerator.scala]: OK, done.

**Comment by JavierJia** [comment]
Looks good! 👍 
The next PR should simply replace the AQLGenerator with SQLGenerator and if the TwitterMap demo works fine, we will switch to SQLPP

**Comment by chenlica** [comment]
Very cool!

**Pull request merged**